reactos/drivers/usb/usbport/device.c

```diff
@@ -0,0 +1,1873 @@
+                                                           ConfigHandle);
+
+                    InterfaceHandle = 0;
+
+                    if (USBD_SUCCESS(USBDStatus))
+                        USBDStatus = USBPORT_OpenInterface(Urb,
+                                                           DeviceHandle,
+                                                           FdoDevice,
+                                                           ConfigHandle,
+                                                           InterfaceInfo,
+                                                           &InterfaceHandle,
+                                                           TRUE);
+
+                    if (InterfaceHandle)
+                    {
+                        InsertTailList(&ConfigHandle->InterfaceHandleList,
+                                       &InterfaceHandle->InterfaceLink);
+                    }
+
+                    if (USBD_ERROR(USBDStatus))
+                        break;
+
+                    ++ix;
+
+                    if (ix >= iNumber)
+                    {
+                        Status = USBPORT_USBDStatusToNtStatus(Urb,
+                                                              USBD_STATUS_SUCCESS);
+                        goto Exit;
+                    }
+
+                    InterfaceInfo = (PUSBD_INTERFACE_INFORMATION)((ULONG_PTR)InterfaceInfo +
+                                                                  InterfaceInfo->Length);
+                }
+
+                Status = USBPORT_USBDStatusToNtStatus(Urb, USBDStatus);
+            }
+            else
+            {
+                Status = USBPORT_USBDStatusToNtStatus(Urb,
+                                                      USBD_STATUS_SET_CONFIG_FAILED);
+            }
+        }
+        else
+        {
+            Status = USBPORT_USBDStatusToNtStatus(Urb,
+                                                  USBD_STATUS_INSUFFICIENT_RESOURCES);
+        }
+    }
+    else
+    {
+        Status = USBPORT_USBDStatusToNtStatus(Urb,
+                                              USBD_STATUS_INVALID_CONFIGURATION_DESCRIPTOR);
+    }
+
+Exit:
+
+    if (NT_SUCCESS(Status))
+    {
+        Urb->UrbSelectConfiguration.ConfigurationHandle = ConfigHandle;
+        DeviceHandle->ConfigHandle = ConfigHandle;
+    }
+    else
+    {
+        DPRINT1("USBPORT_SelectConfiguration: Status %x\n", Status);
+    }
+
+    KeReleaseSemaphore(&FdoExtension->DeviceSemaphore,
+                       LOW_REALTIME_PRIORITY,
+                       1,
+                       FALSE);
+
+    return Status;
+}
+
+VOID
+NTAPI
+USBPORT_AddDeviceHandle(IN PDEVICE_OBJECT FdoDevice,
+                        IN PUSBPORT_DEVICE_HANDLE DeviceHandle)
+{
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+
+    DPRINT("USBPORT_AddDeviceHandle: ... \n");
+
+    FdoExtension = (PUSBPORT_DEVICE_EXTENSION)(FdoDevice->DeviceExtension);
```

Comment:
Deleted all casts for FdoExtension, PdoExtension and "PdoDevice = (PDEVICE_OBJECT)BusContext;" in iface.c